src/utils.ts

```diff
@@ -898,97 +898,224 @@ export function deepCopy<T>(value: T): T {
 }
 
 /** @internal */
-const kBuffers = Symbol('buffers');
-/** @internal */
-const kLength = Symbol('length');
+type ListNode<T> = { value: T; next: ListNode<T>; prev: ListNode<T> };
+/**
+ * T extends NonNullable<unknown> = NonNullable<unknown>
+ * @internal
+ * */
+export class List<T = unknown> {
+  private readonly head: ListNode<T>;
+  private count: number;
+
+  get length() {
+    return this.count;
+  }
+
+  get [Symbol.toStringTag]() {
+    return 'List' as const;
+  }
+
+  constructor() {
+    this.count = 0;
+
+    const sentinel = {
+      value: null as unknown as T,
+      next: null as unknown as ListNode<T>,
+      prev: null as unknown as ListNode<T>
+    };
+    sentinel.next = sentinel as unknown as ListNode<T>;
+    sentinel.prev = sentinel as unknown as ListNode<T>;
+
+    this.head = sentinel;
+  }
+
+  toArray() {
+    return Array.from(this);
+  }
+
+  toString() {
+    return `head <=> ${this.toArray().join(' <=> ')} <=> head`;
+  }
+
+  *[Symbol.iterator](): Generator<T, void, void> {
+    for (const node of this.nodes()) {
+      yield node.value;
+    }
+  }
+
+  private *nodes() {
+    let ptr = this.head.next;
+    while (ptr !== this.head) {
+      // Save next before yielding so that we make removing within iteration safe
+      const { next } = ptr;
+      yield ptr;
+      ptr = next;
+    }
+  }
+
+  /** Insert at end of list */
+  push(value: T) {
+    this.count += 1;
+    const newNode: ListNode<T> = {
+      next: this.head as unknown as ListNode<T>,
+      prev: this.head.prev,
+      value
+    };
+    this.head.prev.next = newNode;
+    this.head.prev = newNode;
+  }
+
+  /** Inserts every item inside an iterable instead of the iterable itself */
+  pushMany(iterable: Iterable<T>) {
+    for (const value of iterable) {
+      this.push(value);
+    }
+  }
+
+  /** Insert at front of list */
+  unshift(value: T) {
+    this.count += 1;
+    const newNode: ListNode<T> = {
+      next: this.head.next,
+      prev: this.head as unknown as ListNode<T>,
+      value
+    };
+    this.head.next.prev = newNode;
+    this.head.next = newNode;
+  }
+
+  private remove(node?: ListNode<T> | null) {
```

Comment:
Let's group all public methods together and all private methods together.  idc where they each go but in the past, I've seen the following structure for classes (top to bottom in a file):

- private variables
- public variables
- getters/setters
- constructor
- private methods
- public methods

whatever order is fine, but let's organize it well 🙂

Reply:
My preference is by public / private because it's easier to understand the public API of the class